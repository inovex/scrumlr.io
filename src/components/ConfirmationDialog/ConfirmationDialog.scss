@import "src/constants/style.scss";

.confirmation-dialog__background {
  position: fixed;
  @include inset-0;
  background-color: rgba($color: $color-backlog-blue, $alpha: 0.1);
  backdrop-filter: blur(10px);
}

.confirmation-dialog__wrapper {
  position: fixed;
  @include inset-0;
  @include flex-center;
}

.confirmation-dialog {
  min-width: 220px;
  max-width: 570px;
  color: $color-black;
  background-color: $color-white;
  border-radius: $rounded--medium;
  padding: 40px 42px 35px 40px;
  border: 2px transparent;
  box-shadow: 0 16px 32px 0 rgba($color-backlog-blue, 0.24);
}

.confirmation-dialog__icon-content-wrapper {
  display: flex;
  width: 100%;
  gap: 16px;
}

.confirmation-dialog__icon {
  min-width: 75px;
  height: 100%;
  margin-top: -9px;
  margin-left: -14px;
  color: var(--accent-color);
  opacity: 0.6;
}

.confirmation-dialog__content {
  width: 100%;
  display: flex;
  flex-direction: column;
  gap: 40px;
}
.confirmation-dialog__title {
  margin: 0px;
  letter-spacing: $letter-spacing--small;
}

.confirmation-dialog__warning {
  display: flex;
  align-items: center;
  justify-content: flex-start;
  gap: 10px;
  margin: 12px 0px 0px 0px;
}
.confirmation-dialog__warning > svg {
  color: var(--accent-color);
  min-width: 14px;
  max-width: 18px;
  opacity: 0.8;
}
.confirmation-dialog__warning > p {
  margin: 0px;
  font-size: $text-size--medium;
  letter-spacing: $letter-spacing--small;
}

@media #{$mini-smartphone} {
  .confirmation-dialog__icon {
    display: none;
  }
  .confirmation-dialog__buttons {
    flex-wrap: wrap;
  }
  .confirmation-dialog__button {
    width: 100%;
  }
  .confirmation-dialog__warning {
    align-items: start;
  }
  .confirmation-dialog {
    width: 80%;
  }
}

.confirmation-dialog__button {
  border: 2px solid $color-backlog-blue;
  border-radius: $rounded--full;
  padding: $padding--small $padding--medium;
  font-size: $text-size--medium;
  font-weight: bold;
  min-height: 48px;
  cursor: pointer;
  transition:
    transform 80ms linear,
    box-shadow 80ms linear;
  &:hover {
    transform: scale(1.05);
    box-shadow: 0 4px 10px 0 rgba($color: $color-backlog-blue, $alpha: 0.2);
  }
}
.confirmation-dialog__button--accept {
  color: $color-white;
  background-color: $color-backlog-blue;
}
.confirmation-dialog__button--decline {
  color: $color-backlog-blue;
  background-color: rgba($color: $color-backlog-blue, $alpha: 0.1);
}
.confirmation-dialog__buttons {
  display: flex;
  justify-content: flex-end;
  gap: $margin--default;
}

.confirmation-dialog__close-button {
  position: absolute;
  top: 0;
  right: 0;
  transform: translate(50%, -50%);
  @include flex-center;
  color: $color-backlog-blue;
  height: $icon--extralarge;
  width: $icon--extralarge;
  border-radius: $rounded--full;
  border: 2px transparent;
  box-shadow: 0 16px 32px 0 rgba($color-backlog-blue, 0.24);
  padding: 0;
  background-color: $color-white;
  cursor: pointer;
  transition:
    color 80ms linear,
    background-color 80ms linear;
  &:hover {
    color: $color-white;
    background-color: $color-backlog-blue;
  }
}
.confirmation-dialog__close-button > svg {
  height: 100%;
  width: 100%;
}
[theme="dark"] {
  .confirmation-dialog__background {
    background-color: rgba($color: $color-planning-pink, $alpha: 0.1);
  }

  .confirmation-dialog {
    color: $color-white;
    background-color: $color-dark-two;
    border-color: $color-planning-pink;
    box-shadow: $box-shadow--dark;
  }

  .confirmation-dialog__button {
    border-color: $color-planning-pink;
    &:hover {
      box-shadow: 0 4px 8px rgba($color: $color-planning-pink, $alpha: 0.2);
    }
  }
  .confirmation-dialog__button--accept {
    background-color: $color-planning-pink;
  }
  .confirmation-dialog__button--decline {
    color: $color-planning-pink;
    background-color: rgba($color: $color-planning-pink, $alpha: 0.1);
  }

  .confirmation-dialog__close-button {
    color: $color-planning-pink;
    background-color: $color-dark-mode-note;
    border-color: $color-planning-pink;
    box-shadow: $box-shadow--dark;
    &:hover {
      color: $color-white;
      background-color: $color-planning-pink;
    }
  }
}
