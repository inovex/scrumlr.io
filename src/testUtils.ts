export type AssertTypeEqual<T, Expected> =
    T extends Expected
        ? (Expected extends T ? true : never)
        : never;