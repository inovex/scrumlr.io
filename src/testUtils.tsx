import {render, RenderOptions} from "@testing-library/react";
import {I18nextProvider} from "react-i18next";
import {FC, PropsWithChildren, ReactElement} from "react";
import {MemoryRouter} from "react-router";
import {Provider} from "react-redux";
import getTestStore from "utils/test/getTestStore";
import i18n from "./i18nTest";

export type AssertTypeEqual<T, Expected> = T extends Expected ? (Expected extends T ? true : never) : never;

const AllTheProvidersWithoutRouter: FC<PropsWithChildren> = ({children}) => (
  <I18nextProvider i18n={i18n}>
    <Provider store={getTestStore()}>{children}</Provider>
  </I18nextProvider>
);

const AllTheProviders: FC<PropsWithChildren> = ({children}) => (
  <MemoryRouter>
    <AllTheProvidersWithoutRouter>{children}</AllTheProvidersWithoutRouter>
  </MemoryRouter>
);

const customRender = (ui: ReactElement, options?: Omit<RenderOptions & {container: Element}, "wrapper">) => render(ui, {wrapper: AllTheProviders, ...options});
const customRenderWithoutRouter = (ui: ReactElement, options?: Omit<RenderOptions & {container: Element}, "wrapper">) =>
  render(ui, {wrapper: AllTheProvidersWithoutRouter, ...options});

export {customRender as render, customRenderWithoutRouter as renderWithoutRouter};
