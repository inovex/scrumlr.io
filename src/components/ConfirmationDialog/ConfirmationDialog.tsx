import {FC} from "react";
import {useTranslation} from "react-i18next";
import {animated, Transition} from "react-spring";
import {dialogTransitionConfig} from "utils/transitionConfig";
import {Portal} from "components/Portal";
import {ReactComponent as CloseIcon} from "assets/icon-close.svg";
import {ReactComponent as WarningIcon} from "assets/icon-warning.svg";
import "./ConfirmationDialog.scss";

type ConfirmationDialogProps = {
  title: string;
  onAccept: () => void;
  onAcceptLabel?: string;
  onDecline: () => void;
  onDeclineLabel?: string;
  onExtraOption?: () => void;
  onExtraOptionLabel?: string;
  icon?: React.FunctionComponent<React.SVGProps<SVGSVGElement>>;
  warning?: boolean;
};

export const ConfirmationDialog: FC<ConfirmationDialogProps> = (props) => {
  const {t} = useTranslation();

  return (
    <Portal onClose={props.onDecline}>
      <div className="confirmation-dialog__background" />
      <div className="confirmation-dialog__wrapper">
        <Transition {...dialogTransitionConfig}>
          {(styles) => (
            <animated.aside aria-modal="true" aria-label={props.title} className="confirmation-dialog" role="dialog" style={styles}>
              <button aria-label="Close dialog" className="confirmation-dialog__close-button" onClick={() => props.onDecline()} type="button">
                <CloseIcon className="dialog__close-icon" />
              </button>
              <div className="confirmation-dialog__icon-content-wrapper">
                {props.icon && <props.icon className="confirmation-dialog__icon" />}
                <div className="confirmation-dialog__content">
                  <div>
                    <h2 className="confirmation-dialog__title">{props.title}</h2>
                    {props.warning && (
                      <div className="confirmation-dialog__warning">
                        <WarningIcon />
                        <p>{t("ConfirmationDialog.warning")}</p>
                      </div>
                    )}
                  </div>
                  <div className="confirmation-dialog__buttons">
                    <button
                      aria-label={props.onDeclineLabel ?? t("ConfirmationDialog.no")}
                      className="confirmation-dialog__button confirmation-dialog__button--decline"
                      onClick={() => props.onDecline()}
                      type="button"
                    >
                      {props.onDeclineLabel ?? t("ConfirmationDialog.no")}
                    </button>
                    <button
                      aria-label={props.onAcceptLabel ?? t("ConfirmationDialog.yes")}
                      className="confirmation-dialog__button confirmation-dialog__button--accept"
                      onClick={() => props.onAccept()}
                      type="button"
                    >
                      {props.onAcceptLabel ?? t("ConfirmationDialog.yes")}
                    </button>
                    {props.onExtraOption && props.onExtraOptionLabel && (
                      <button
                        aria-label={props.onExtraOptionLabel}
                        className="confirmation-dialog__button confirmation-dialog__button--decline"
                        onClick={() => props.onExtraOption!()}
                        type="button"
                      >
                        {props.onExtraOptionLabel}
                      </button>
                    )}
                  </div>
                </div>
              </div>
            </animated.aside>
          )}
        </Transition>
      </div>
    </Portal>
  );
};
