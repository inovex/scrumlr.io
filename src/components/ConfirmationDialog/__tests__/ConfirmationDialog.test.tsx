import {render} from "testUtils";
import {ConfirmationDialog} from "../ConfirmationDialog";

describe("<ConfirmationDialog />", () => {
  beforeEach(() => {
    const portal = global.document.createElement("div");
    portal.setAttribute("id", "portal");
    global.document.querySelector("body")!.appendChild(portal);
  });

  test("snapshot test", () => {
    const {container} = render(<ConfirmationDialog title="headline" onAcceptLabel="accept" onDeclineLabel="decline" onAccept={jest.fn()} onDecline={jest.fn()} />, {
      container: global.document.querySelector("#portal")!,
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
