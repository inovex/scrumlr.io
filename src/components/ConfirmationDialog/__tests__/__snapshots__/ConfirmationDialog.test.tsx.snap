// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConfirmationDialog /> snapshot test 1`] = `
<div
  class="portal"
  role="dialog"
>
  <div
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="0"
  />
  <div
    data-focus-lock-disabled="false"
  >
    <div
      class="portal__frame accent-color__backlog-blue"
    >
      <div
        class="portal__content"
        role="dialog"
      >
        <div
          class="confirmation-dialog__background"
        />
        <div
          class="confirmation-dialog__wrapper"
        >
          <aside
            aria-label="headline"
            aria-modal="true"
            class="confirmation-dialog"
            role="dialog"
            style="opacity: 0; transform: translateY(-20px);"
          >
            <button
              aria-label="Close dialog"
              class="confirmation-dialog__close-button"
              type="button"
            >
              <svg
                class="dialog__close-icon"
              >
                icon-close.svg
              </svg>
            </button>
            <div
              class="confirmation-dialog__icon-content-wrapper"
            >
              <div
                class="confirmation-dialog__content"
              >
                <div>
                  <h2
                    class="confirmation-dialog__title"
                  >
                    headline
                  </h2>
                </div>
                <div
                  class="confirmation-dialog__buttons"
                >
                  <button
                    aria-label="decline"
                    class="confirmation-dialog__button confirmation-dialog__button--decline"
                    type="button"
                  >
                    decline
                  </button>
                  <button
                    aria-label="accept"
                    class="confirmation-dialog__button confirmation-dialog__button--accept"
                    type="button"
                  >
                    accept
                  </button>
                </div>
              </div>
            </div>
          </aside>
        </div>
      </div>
    </div>
  </div>
  <div
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="0"
  />
</div>
`;
